starlette_exporter/middleware.py

```diff
@@ -49,10 +49,24 @@ async def dispatch(self, request, call_next):
 
         finally:
             # group_paths enables returning the original router path (with url param names)
-            # the second check is to ensure that an endpoint was matched before trying to determine the name.
+            # for example, when using this option, requests to /api/product/1 and /api/product/3
+            # will both be grouped under /api/product/{product_id}. See the README for more info.
             if self.group_paths and request.scope.get('endpoint', None) and request.scope.get('router', None):
+
                 try:
-                    path = [route for route in request.scope['router'].routes if route.endpoint == request.scope['endpoint']][0].path
+                    # try to match the request scope's handler function against one of handlers in the app's router.
+                    # if a match is found, return the path used to mount the handler (e.g. api/product/{product_id}).
+                    path = [
+                        route for route in request.scope['router'].routes
+                            if (hasattr(route, 'endpoint') and route.endpoint  == request.scope['endpoint'])
```

Comment:
Since you're targeting `starlette` here and `starlette` is pretty well typed, I think an `isinstance` check could also be used to check for the cases and would make it clearer to the reader what is actually happening.

Reply:
Thanks for looking the commit over... I'll keep your suggestion in mind and try to work on refactoring it a bit in the near future. The only issue I ran into with `isinstance` is that checking the type isn't strictly necessary just to match against a route in the router (unless in the future there's a need to handle `Mount` objects separately... right now I think we just need to be able to match it up with a route). If you disagree I would welcome a PR with a change.

There might also be an easier and more clear way to group the metrics by endpoint... I just haven't found it yet.